ThirdPartyModules/Amasty/Rewards.php

```diff
@@ -0,0 +1,106 @@
+<?php
+/**
+ * Bolt magento2 plugin
+ *
+ * NOTICE OF LICENSE
+ *
+ * This source file is subject to the Open Software License (OSL 3.0)
+ * that is bundled with this package in the file LICENSE.txt.
+ * It is also available through the world-wide-web at this URL:
+ * http://opensource.org/licenses/osl-3.0.php
+ *
+ * @category   Bolt
+ * @package    Bolt_Boltpay
+ * @copyright  Copyright (c) 2017-2020 Bolt Financial, Inc (https://www.bolt.com)
+ * @license    http://opensource.org/licenses/osl-3.0.php  Open Software License (OSL 3.0)
+ */
+
+namespace Bolt\Boltpay\ThirdPartyModules\Amasty;
+
+use Bolt\Boltpay\Helper\Bugsnag;
+use Bolt\Boltpay\Helper\Discount;
+use Bolt\Boltpay\Helper\Shared\CurrencyUtils;
+
+class Rewards
+{
+    /**
+     * @var Bugsnag
+     */
+    private $bugsnagHelper;
+    
+    /**
+     * @var Discount
+     */
+    protected $discountHelper;
+
+    /**
+     * @param Bugsnag $bugsnagHelper Bugsnag helper instance
+     * @param Discount $discountHelper
+     */
+    public function __construct(
+        Bugsnag   $bugsnagHelper,
+        Discount  $discountHelper
+    ) {
+        $this->bugsnagHelper   = $bugsnagHelper;
+        $this->discountHelper  = $discountHelper;
+    }
+
+    /**
+     * @param array $result
+     * @param Amasty\Rewards\Helper\Data $amastyRewardsHelperData
+     * @param Quote $quote
+     * 
+     * @return array
+     */
+    public function collectDiscounts($result,
+                                     $amastyRewardsHelperData,
+                                     $quote,
+                                     $parentQuote,
+                                     $paymentOnly)
+    {
+        list ($discounts, $totalAmount, $diff) = $result;
+
+        try {
+            if ($quote->getData('amrewards_point')) {
+                $rewardData = $amastyRewardsHelperData->getRewardsData();
+                $pointsUsed = $rewardData['pointsUsed'];
+                $pointsRate = $rewardData['rateForCurrency'];
+                $amount = $pointsUsed / $pointsRate;
+                $currencyCode = $quote->getQuoteCurrencyCode();
+                $roundedAmount = CurrencyUtils::toMinor($amount, $currencyCode);
+                $newDiscounts = array();
+                // Amasty Rewards plugin applies discount amount to the quote as same as coupon code,
+                // so we need to collect the amount discounted with reward points separately. 
+                foreach ($discounts as $discount) {
```

Comment:
@dumega can you explain the logic in this loop a bit better?
Why do we need to recreate the whole discount array instead of just adding to it?
Seems like we can potentially lose the internal coupon discount here, if applied, but I may be wrong.
Did you test with both Magento coupon and Amasty Reward applied?

Reply:
@danac-gs Actually that's the purpose of this PR, let me show you some screenshots

![image](https://user-images.githubusercontent.com/3198527/93284388-768fc280-f787-11ea-973c-9ceea8f34288.png)
![image](https://user-images.githubusercontent.com/3198527/93284395-7c85a380-f787-11ea-862e-9896e2805c4f.png)

As you can see, the Amasty reward points combine with Magento coupon, and when the Bolt plugin try to collect discount amount from address object, it just get the reward points and Magento coupon as a whole, and then sent to the Bolt server as coupon discount. So this PR is to separate Amasty reward points from Magento coupon.

![image](https://user-images.githubusercontent.com/3198527/93284908-8cea4e00-f788-11ea-80a3-2054b6e38a50.png)

Reply:
Great, thanks.